pkg/cloud-init/cloud-init.go

```diff
@@ -0,0 +1,349 @@
+}
+
+func fileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	exists := false
+
+	if err == nil {
+		exists = true
+	} else if os.IsNotExist(err) {
+		err = nil
+	}
+	return exists, err
+}
+
+func md5CheckSum(filePath string) ([]byte, error) {
+	var result []byte
+
+	file, err := os.Open(filePath)
+	if err != nil {
+		return result, err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	_, err = io.Copy(hash, file)
+
+	if err != nil {
+		return result, err
+	}
+
+	result = hash.Sum(result)
+	return result, nil
+}
+
+func setFileOwnership(username string, file string) error {
+	usrObj, err := user.Lookup(username)
+	if err != nil {
+		return err
+	}
+
+	uid, err := strconv.Atoi(usrObj.Uid)
+	if err != nil {
+		return err
+	}
+
+	gid, err := strconv.Atoi(usrObj.Gid)
+	if err != nil {
+		return err
+	}
+
+	return os.Chown(file, uid, gid)
+}
+
+func filesAreEqual(path1 string, path2 string) (bool, error) {
+	exists, err := fileExists(path1)
+	if err != nil {
+		return false, err
+	} else if exists == false {
+		return false, nil
+	}
+
+	exists, err = fileExists(path2)
+	if err != nil {
+		return false, err
+	} else if exists == false {
+		return false, nil
+	}
+
+	sum1, err := md5CheckSum(path1)
+	if err != nil {
+		return false, err
+	}
+	sum2, err := md5CheckSum(path2)
+	if err != nil {
+		return false, err
+	}
+
+	return bytes.Equal(sum1, sum2), nil
+}
+
+// The unit test suite uses this function
+func SetIsoCreationFunction(isoFunc IsoCreationFunc) {
+	cloudInitIsoFunc = isoFunc
+}
+
+// The unit test suite uses this function
+func SetLocalDataOwner(user string) {
+	cloudInitOwner = user
+}
+
+func SetLocalDirectory(dir string) error {
+
+	err := os.MkdirAll(dir, 0755)
+	if err != nil {
+		return errors.New(fmt.Sprintf("Unable to initalize cloudInit local cache directory (%s). %v", dir, err))
+	}
+
+	exists, err := fileExists(dir)
+	if err != nil {
+		return errors.New(fmt.Sprintf("CloudInit local cache directory (%s) does not exist or is inaccessible. %v", dir, err))
+	} else if exists == false {
+		return errors.New(fmt.Sprintf("CloudInit local cache directory (%s) does not exist or is inaccessible.", dir))
+	}
+
+	cloudInitLocalDir = dir
+	return nil
+}
+
+func GetDomainBasePath(domain string, namespace string) string {
+	return fmt.Sprintf("%s/%s/%s", cloudInitLocalDir, namespace, domain)
+}
+
+// This is called right before a VM is defined with libvirt.
+// If the cloud-init type requires altering the domain, this
+// is the place to do that.
+func InjectDomainData(vm *v1.VM) (*v1.VM, error) {
+	namespace := precond.MustNotBeEmpty(vm.GetObjectMeta().GetNamespace())
+	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
+	if vm.Spec.CloudInit == nil {
+		return vm, nil
+	}
+
+	err := ValidateArgs(vm.Spec.CloudInit)
+	if err != nil {
+		return vm, err
+	}
+
+	switch vm.Spec.CloudInit.DataSource {
+	case dataSourceNoCloud:
+		filePath := fmt.Sprintf("%s/%s", GetDomainBasePath(domain, namespace), noCloudFile)
```

Comment:
Security risk as domain is pretty unfiltered user-input. You might want to use the uuid instead.

Reply:
domain and namespace shouldn't be allowed to be structured in anyway that would present a security risk here.

As a safety precaution we could return an error if any characters are present that are not alphanumeric, '.' or '-'. Those are the only characters that should ever be in the namespace or domain, so under normal operation that safety check should never get triggered.

Reply:
I hesitate to set a restriction on the name of a domain, because we are using it to create some paths.

We have alternatives which don't suffer from this: Some kind of encoding or using the uuid (which is effectively a charset limited name :) )

Reply:
> I hesitate to set a restriction on the name of a domain, because we are using it to create some paths.

We aren't making up the restrictions to adhere to path creation. These restrictions already exist (domain name has to adhere to pod naming restrictions) and happen to be path safe.

Reply:
here some documentation from k8s regarding to allowed characters: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/

btw. also keep in mind that uuids are spatially and temporally unique. In some cases that could lead to a situation where we loose track of a resource on a VM repost (don't know if that would be an issue here).

Reply:
from the link: "By convention, the names of Kubernetes resources should be ...", that's not really a precise limitation, and in the linked https://git.k8s.io/community/contributors/design-proposals/identifiers.md there is no such limitation mentioned nor enforced.

I would like to avoid situation where we choose name to be the unique identifier internally and externally. Users have a tendency to rename "stuff" and kind of expecting the object remain the same.

Reply:
Actually it is enforced on TPRs. We can take the UUIDs in that case (since we use the same UID for the Domain in libvirt like in the k8s cluster). However, that is not always possible. Also here doing clean-ups after a virt-handler resync becomes more difficult, but solvable.

Just to provide more context, since I am not sure if that is clear for people: 

The **primary keys** are always namespace and name. It is not possible in general to query objects by UID in k8s (it might be possible in the future though). Namespace and name are always also internal identifiers. In some cases it matters if the UUID is different (the actual object is a repost), in others not. 

An example where you would not care about different UUIDs: Secrets are referenced on the Pod by name, and when a container starts, it is pulled in by name. If the container is restarted, it will be pulled in by name again, not by UUID. 

An example where a controller cares: In our case, when we re-post a VM, we need to make sure that we construct an appropriate Pod for it. Thus, we need to make sure that the pod on the node really belongs to the VM instance with a specific UUID. Therefore one of our controllers internally tracks the UUIDs too.

An end-user can normally not reference other objects by UUID. Controllers sometimes do that internally.

Reply:
> Users have a tendency to rename "stuff" and kind of expecting the object remain the same.
the domain name can not change without resulting in a completely new VM.